File: apps/worker/src/routes/analytics.ts
import { Hono } from 'hono'
import type { Env } from '../index'
import { getPhone, getUser } from '../middleware/auth'
import { getSupabase } from '../lib/supabase'

const app = new Hono<{ Bindings: Env }>()

app.get('/builds', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const sb = getSupabase(c.env)
  const since = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString()
  const { data } = await sb.from('linq_jobs').select('status,duration_ms').eq('user_phone', phone).gte('created_at', since)
  const jobs = data ?? []
  const total = jobs.length
  const complete = jobs.filter(j => j.status === 'complete').length
  const failed = jobs.filter(j => j.status === 'failed').length
  const durations = jobs.map(j => j.duration_ms).filter(Boolean) as number[]
  const avg_duration_ms = durations.length ? Math.round(durations.reduce((a, b) => a + b, 0) / durations.length) : null
  return c.json({ total, complete, failed, avg_duration_ms })
})

app.get('/usage', async (c) => {
  let phone: string
  try { phone = await getPhone(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const sb = getSupabase(c.env)
  const now = new Date()
  const periods = Array.from({ length: 3 }, (_, i) => {
    const d = new Date(now.getFullYear(), now.getMonth() - i, 1)
    return `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, '0')}`
  })
  const { data } = await sb.from('linq_usage').select('*').eq('user_phone', phone).in('billing_period', periods).order('billing_period', { ascending: false })
  return c.json(data ?? [])
})

app.get('/revenue', async (c) => {
  let user: Record<string, unknown>
  try { user = await getUser(c) } catch (e) { return c.json({ error: String(e) }, 401) }
  const sb = getSupabase(c.env)
  const phone = user.phone_number as string
  const { data: tenant } = await sb.from('nano_tenants').select('id,current_balance_cents').eq('owner_phone', phone).maybeSingle()
  if (!tenant) return c.json({ error: 'Tenant not found' }, 404)
  const { data: rows } = await sb.from('nano_ledger').select('type,amount_cents').eq('tenant_id', tenant.id)
  const ledger = rows ?? []
  const earned = ledger.filter(r => r.type === 'credit').reduce((a, r) => a + r.amount_cents, 0)
  const paid_out = ledger.filter(r => r.type === 'debit').reduce((a, r) => a + r.amount_cents, 0)
  return c.json({ total_earned: earned, total_paid_out: paid_out, current_balance: tenant.current_balance_cents ?? 0 })
})

export default app

File: apps/api/app/routers/analytics.py
import datetime

from fastapi import APIRouter, Depends, HTTPException

from app.auth import get_current_user, get_current_user_phone
from app.deps import optional_supabase

router = APIRouter(prefix="/analytics", tags=["analytics"])


@router.get("/builds")
async def build_analytics(
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    since = (datetime.datetime.utcnow() - datetime.timedelta(days=30)).isoformat()
    row = (
        db.table("linq_jobs")
        .select("status,duration_ms")
        .eq("user_phone", phone)
        .gte("created_at", since)
        .execute()
    )
    jobs = row.data or []
    total = len(jobs)
    complete = sum(1 for j in jobs if j.get("status") == "complete")
    failed = sum(1 for j in jobs if j.get("status") == "failed")
    durations = [j["duration_ms"] for j in jobs if j.get("duration_ms")]
    avg_ms = int(sum(durations) / len(durations)) if durations else None
    return {"total": total, "complete": complete, "failed": failed, "avg_duration_ms": avg_ms}


@router.get("/usage")
async def usage_analytics(
    phone: str = Depends(get_current_user_phone),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    now = datetime.datetime.utcnow()
    periods = []
    for i in range(3):
        m, y = now.month - i, now.year
        if m <= 0:
            m += 12
            y -= 1
        periods.append(f"{y}-{m:02d}")
    row = (
        db.table("linq_usage")
        .select("*")
        .eq("user_phone", phone)
        .in_("billing_period", periods)
        .order("billing_period", desc=True)
        .execute()
    )
    return row.data


@router.get("/revenue")
async def revenue_analytics(
    user: dict = Depends(get_current_user),
    db=Depends(optional_supabase),
):
    if not db:
        raise HTTPException(503, "DB unavailable")
    tenant = (
        db.table("nano_tenants")
        .select("id,current_balance_cents")
        .eq("owner_phone", user["phone_number"])
        .maybe_single()
        .execute()
    )
    if not tenant.data:
        raise HTTPException(404, "Tenant not found")
    tid = tenant.data["id"]
    ledger = (
        db.table("nano_ledger")
        .select("type,amount_cents")
        .eq("tenant_id", tid)
        .execute()
    )
    rows = ledger.data or []
    earned = sum(r["amount_cents"] for r in rows if r.get("type") == "credit")
    paid_out = sum(r["amount_cents"] for r in rows if r.get("type") == "debit")
    return {
        "total_earned": earned,
        "total_paid_out": paid_out,
        "current_balance": tenant.data.get("current_balance_cents", 0),
    }
